### DISCUSSION TITLE
Configuring XPRESS solver path

### DISCUSSION THREAD

**Discussion by ScJa** [coin-or/pulp#143] [Q&A]
Hi, I'm having trouble hooking up my XPRESS-solver with PuLP.

I'm working on Windows 10 with PuLP 1.6 and had the FICO XPRESS solver installed already before installing PuLP.

When I run `pulp.pulpTestAll()` it doesn't show the XPRESS solver as unavailable but instead the test fails with the message:

```
The system cannot find the path specified. 
pulp.constants.PulpError: Tests Failed
```

I have found the pulp.cfg.win file and tried entering my path in some different ways (e.g. XpressPath = C:\xpressmp\bin\optimizer.exe, but it didn't change the error.

How do I have to specify the path?

**Comment by stumitchell** [comment]
Hi I actually don't have a copy of xpress to test with but I will see if I can get hold of a trail version and test this.

Stu

**Comment by ScJa** [comment]
Hi Stuart,
thanks for taking a look at it. If you can't get a hold of one I can some settings on my machine if you want to. 

I realized I missed one line in the error statement:
`PuLP: Error while executing optimizer.exe`

Furthermore I've tried to debug PuLP to see if it's a trivial issue. It seems to me as the if the path gets picked up correctly at least and is saved within the XPRESS-solver class.
Maybe this helps,

Jakob

**Comment by stumitchell** [comment]
you can try

prob.solve(XPRESS(path= "C:\xpressmp\bin\optimizer.exe"))

see if that works


On Mon, Jul 24, 2017 at 9:24 PM, Schneider Jakob <notifications@github.com>
wrote:

> Hi Stuart,
> thanks for taking a look at it. If you can't get a hold of one I can some
> settings on my machine if you want to.
>
> I realized I missed one line in the error statement:
> PuLP: Error while executing optimizer.exe
>
> Furthermore I've tried to debug PuLP to see if it's a trivial issue. It
> seems to me as the if the path gets picked up correctly at least and is
> saved within the XPRESS-solver class.
> Maybe this helps,
>
> Jakob
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/coin-or/pulp/issues/143#issuecomment-317367659>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAtEWUsQXGjCtP6U39GiZSKVUPWJxoqnks5sRGLggaJpZM4OgKwV>
> .
>



-- 
Stuart Mitchell
PhD Engineering Science
Extraordinary Freelance Programmer and Optimisation Guru
www.stuartmitchell.com

**Comment by ScJa** [comment]
Hi,

I got a similar error message:
```
The system cannot find the path specified.
OSError: [Errno 22] Invalid argument

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "my/path", line 105, in myfile
    rdl_model.solve(XPRESS(path="C:\\xpressmp\\bin\\optimizer.exe"))
  File "C:\Python36-64\lib\site-packages\pulp\pulp.py", line 1664, in solve
    status = solver.actualSolve(self, **kwargs)
  File "C:\Python36-64\lib\site-packages\pulp\solvers.py", line 1286, in actualSolve
    if xpress.close() != None:
  File "C:\Python36-64\lib\os.py", line 993, in close
    self._stream.close()
OSError: [Errno 22] Invalid argument
```

**Comment by stumitchell** [comment]
Can you try with python 2.7??

On Tue, Jul 25, 2017 at 9:03 PM, Schneider Jakob <notifications@github.com>
wrote:

> Hi,
>
> I got a similar error message:
>
> The system cannot find the path specified.
> OSError: [Errno 22] Invalid argument
>
> During handling of the above exception, another exception occurred:
>
> Traceback (most recent call last):
>   File "my/path", line 105, in myfile
>     rdl_model.solve(XPRESS(path="C:\\xpressmp\\bin\\optimizer.exe"))
>   File "C:\Python36-64\lib\site-packages\pulp\pulp.py", line 1664, in solve
>     status = solver.actualSolve(self, **kwargs)
>   File "C:\Python36-64\lib\site-packages\pulp\solvers.py", line 1286, in actualSolve
>     if xpress.close() != None:
>   File "C:\Python36-64\lib\os.py", line 993, in close
>     self._stream.close()
> OSError: [Errno 22] Invalid argument
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/coin-or/pulp/issues/143#issuecomment-317676030>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAtEWYXi5rUcfXq3JZN2-wAnTpJt4br6ks5sRa9cgaJpZM4OgKwV>
> .
>



-- 
Stuart Mitchell
PhD Engineering Science
Extraordinary Freelance Programmer and Optimisation Guru
www.stuartmitchell.com

**Comment by ScJa** [comment]
@Gives me a similar error when I run `prob.solve(XPRESS(path= "C:\xpressmp\bin\optimizer.exe"))`

```
>>> rdl_model.solve(XPRESS(path= "C:\\xpressmp\\bin\\optimizer.exe"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\schnjako\p27\lib\site-packages\pulp\pulp.py", line 1664, in solve
    status = solver.actualSolve(self, **kwargs)
  File "C:\Users\schnjako\p27\lib\site-packages\pulp\solvers.py", line 1287, in actualSolve
    raise PulpSolverError("PuLP: Error while executing "+self.path)
pulp.solvers.PulpSolverError: PuLP: Error while executing C:\xpressmp\bin\optimizer.exe
````
`pulpTestAll()` fails with 
```
PuLP: Error while executing optimizer.exe
('* Solver', <class pulp.solvers.XPRESS at 0x02209928>, 'failed.')
``` 

For now I will probably look into the Xpress Python interface.

**Comment by wzxqwe** [comment]
Hello, I'm having the same error "PulpSolverError: PuLP: Error while executing optimizer.exe" 
For my situation i really want to stick in pulp, is there any suggestion?

**Comment by SanPen** [comment]
Hi all, 

I would like to fix the Xpress interface of PuLP.

So far I have pinpointed the problem;

I run the very simple example Whiskas model 1.

In the pulp file `solvers.py`, in the class `XPRESS` and the function `actualSolve`, I found that the command-line executable `optimizer.exe`  operated via `os.popen` is not responding as expected.

The process is to save the problem as a .lp file, make the solver ( `optimizer.exe`) read the file, set some parameters in the solver, minimize / maximize the problem, make the solver save a .prt file and parse that file back into PuLP.

The problem seems to be that somehow the instructions are not reaching the solver.
I concluded that because, if I call manually optimizer.exe from the command line and perform the same instructions as python tries to set into the solver everything works, but under python it does not work.

So at this point I'll try to play around with `os.popen` and `subprocess.Popen` to see if I get it fixed.

@stumitchell any suggestions are very welcome.

**Comment by pchtsp** [comment]
you could check how the solve_CBC function works in COIN_CMD class or the
actualSolve function works in both CPLEX_CMD and GUROBI_CMD classes. All
three work correctly and are very similar in the steps they use to build
the call to the executable.

**Comment by SanPen** [comment]
Hi, I think I solved it.

The problem is that the xpress command line application was unable to interpret systems paths.
So if I remove the path and just set the .lp file name to `file.lp` instaead of `c:\folder1\folder2\folder3\file.lp` it works as expected.

```
    def actualSolve(self, lp):
        """
        Solve a well formulated lp problem
        """
        if not self.executable(self.path):
            raise PulpSolverError("PuLP: cannot execute " + self.path)

        # if not self.keepFiles:
        #     uuid = uuid4().hex
        #     tmpLp = os.path.join(self.tmpDir, "%s-pulp.lp" % uuid)
        #     tmpSol = os.path.join(self.tmpDir, "%s-pulp.prt" % uuid)
        # else:
        #     tmpLp = lp.name+"-pulp.lp"
        #     tmpSol = lp.name+"-pulp.prt"

        if not self.keepFiles:

            # the xpress solver has issues reading the system paths (uncomment to see this)
            # uuid = uuid4().hex
            # tmpLp = os.path.join(self.tmpDir, "%s-pulp.lp" % uuid)
            # tmpSol = os.path.join(self.tmpDir, "%s-pulp.prt" % uuid)
            tmpLp = lp.name.replace(' ', '_') + "-pulp.lp"
            tmpSol = lp.name.replace(' ', '_') + "-pulp.prt"
        else:
            tmpLp = lp.name.replace(' ', '_') + "-pulp.lp"
            tmpSol = lp.name.replace(' ', '_') + "-pulp.prt"

        lp.writeLP(tmpLp, writeSOS=1, mip=self.mip)

        if not self.msg:
            xpress = os.popen(self.path+" "+lp.name+" > /dev/null 2> /dev/null", "w")
        else:
            xpress = os.popen(self.path+" "+lp.name, "w")

        xpress.write("READPROB " + tmpLp + "\n")

        if self.maxSeconds:
            xpress.write("MAXTIME=%d\n" % self.maxSeconds)

        if self.targetGap:
            xpress.write("MIPRELSTOP=%f\n" % self.targetGap)

        if self.heurFreq:
            xpress.write("HEURFREQ=%d\n" % self.heurFreq)

        if self.heurStra:
            xpress.write("HEURSTRATEGY=%d\n" % self.heurStra)

        if self.coverCuts:
            xpress.write("COVERCUTS=%d\n" % self.coverCuts)

        if self.preSolve:
            xpress.write("PRESOLVE=%d\n" % self.preSolve)

        for option in self.options:
            xpress.write(option+"\n")

        if lp.sense == LpMaximize:
            xpress.write("MAXIM\n")
        else:
            xpress.write("MINIM\n")

        if lp.isMIP() and self.mip:
            xpress.write("GLOBAL\n")
        xpress.write("WRITEPRTSOL " + tmpSol + "\n")
        xpress.write("QUIT\n")

        if xpress.close() is not None:
            raise PulpSolverError("PuLP: Error while executing " + self.path)

        status, values = self.readsol(tmpSol)

        if not self.keepFiles:
            try:
                os.remove(tmpLp)
            except:
                pass
            try:
                os.remove(tmpSol)
            except:
                pass
        lp.status = status
        lp.assignVarsVals(values)
        if abs(lp.infeasibilityGap(self.mip)) > 1e-5: # Arbitrary
            lp.status = LpStatusInfeasible
        return lp.status
```

To test this, you can download the xpress community edition for small problems (I think up to 1000 variables) they provide executables for windows and OSX. For the paid version that I use at work they provide linux versions too.

**Comment by s-c-e** [comment]
Hi all,

I've been using pulp with Xpress on Windows lately. Here's what I've observed and the work-arounds I've been using, if it helps others:

1) It doesn't play well with "empty" constraints. I could pass something like "4 <= lpSum( [] )" when using CBC, but not with Xpress. This comes up in practice if you were to be building constraints across a series, and some of those series don't have data. The work-around for Xpress has been to evaluate the contents going in to lpSum, and if the length is non-zero, to build the constraint. Both the RHS and LHS must contain something - again only odd because CBC is able to put a zero in its place?

2) When passing the problem to Xpress, I have to make sure the problem name does not contain spaces. Not really a big problem, it just has to do with how optimizer.exe is called.

3) When passing the problem to Xpress, I have to make sure msg is True and keepFiles is True ```prob.solve(pulp.XPRESS(msg=True, keepFiles=True))```

4) Infeasibility or no-solution-found handling is a bit rough. Instead of with CBC when you get the problem status back as "not found" or "infeasible", I have to do a bit of extra handling like this:
```
try:
    prob.solve(pulp.XPRESS(msg=True, keepFiles=True,))
except IndexError:
    prob.status = pulp.LpStatusNotSolved
```

**Comment by pchtsp** [comment]
@s-c-e thanks would you care making a PR with some of those workarounds? for example, for number 2 we now take out empty spaces and show a warning. Maybe we could do other checks for `keepFiles` and empty constraints ?
